From 8b4603614f4d3a9531e4cdffcf8062688186f848 Mon Sep 17 00:00:00 2001
From: Hou Tao <houtao1@huawei.com>
Date: Wed, 15 Mar 2023 17:37:43 +0800
Subject: [PATCH 1607/1646] [Huawei] scsi: cancel the inflight async device
 probe when remove scsi_target

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188150

----------------------------------------

Cancel the inflight async device probe when removing scsi_target, so
no new disk will be added when __scsi_target_remove() returns.

Signed-off-by: Hou Tao <houtao1@huawei.com>
Signed-off-by: Zhong Jinghua <zhongjinghua@huawei.com>
Reviewed-by: Hou Tao <houtao1@huawei.com>
Signed-off-by: Jialin Zhang <zhangjialin11@huawei.com>
---
 drivers/scsi/scsi_sysfs.c | 36 ++++++++++++++++++++++++++++++++++++
 1 file changed, 36 insertions(+)

diff --git a/drivers/scsi/scsi_sysfs.c b/drivers/scsi/scsi_sysfs.c
index 9b80a084e434..e15f561e756c 100644
--- a/drivers/scsi/scsi_sysfs.c
+++ b/drivers/scsi/scsi_sysfs.c
@@ -1511,6 +1511,40 @@ void scsi_remove_device(struct scsi_device *sdev)
 }
 EXPORT_SYMBOL(scsi_remove_device);
 
+/* Cancel the inflight async probe for scsi_device */
+static void __scsi_kill_devices(struct scsi_target *starget)
+{
+	struct Scsi_Host *shost = dev_to_shost(starget->dev.parent);
+	struct scsi_device *sdev, *to_put = NULL;
+	unsigned long flags;
+
+	spin_lock_irqsave(shost->host_lock, flags);
+	list_for_each_entry(sdev, &shost->__devices, siblings) {
+		if (sdev->channel != starget->channel ||
+		    sdev->id != starget->id)
+			continue;
+
+		if ((sdev->sdev_state != SDEV_DEL &&
+		     sdev->sdev_state != SDEV_CANCEL) || !sdev->is_visible)
+			continue;
+		if (!kobject_get_unless_zero(&sdev->sdev_gendev.kobj))
+			continue;
+		spin_unlock_irqrestore(shost->host_lock, flags);
+
+		if (to_put)
+			put_device(&to_put->sdev_gendev);
+		device_lock(&sdev->sdev_gendev);
+		kill_device(&sdev->sdev_gendev);
+		device_unlock(&sdev->sdev_gendev);
+		to_put = sdev;
+
+		spin_lock_irqsave(shost->host_lock, flags);
+	}
+	spin_unlock_irqrestore(shost->host_lock, flags);
+	if (to_put)
+		put_device(&to_put->sdev_gendev);
+}
+
 static void __scsi_remove_target(struct scsi_target *starget)
 {
 	struct Scsi_Host *shost = dev_to_shost(starget->dev.parent);
@@ -1540,6 +1574,8 @@ static void __scsi_remove_target(struct scsi_target *starget)
 		goto restart;
 	}
 	spin_unlock_irqrestore(shost->host_lock, flags);
+
+	__scsi_kill_devices(starget);
 }
 
 /**
-- 
2.17.1

